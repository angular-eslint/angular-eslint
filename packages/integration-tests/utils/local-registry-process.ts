/* eslint-disable @typescript-eslint/no-non-null-assertion */
import type { ChildProcess } from 'child_process';
import { spawn } from 'child_process';
import execa from 'execa';
import { join } from 'path';
import kill from 'tree-kill';

export const FIXTURES_DIR = join(__dirname, '../fixtures/');

// Used to ensure all the time-consuming setup steps for fixtures do not cause jest to time out
export const LONG_TIMEOUT_MS = 600000; // 10 mins

let localRegistryProcess: ChildProcess;

const VERSION = `9999.0.1-local-integration-tests`;

const cwd = process.cwd();

process.env.npm_config_registry = `http://localhost:4872/`;
process.env.YARN_REGISTRY = process.env.npm_config_registry;

export async function spawnLocalRegistry(): Promise<void> {
  localRegistryProcess = spawn('npx', [
    'verdaccio',
    '--config',
    './local-registry/config.yml',
    '--listen',
    '4872',
  ]);

  const collectedOutput: string[] = [];
  let resolvedOrRejected = false;

  setTimeout(() => {
    if (!resolvedOrRejected) {
      console.error(`Failed to start the npm registry`);
      console.error(collectedOutput.join(''));
      cleanUp(1);
    }
  }, 10000);

  await new Promise((res, rej) => {
    localRegistryProcess.stdout!.pipe(process.stdout);
    localRegistryProcess.stdout!.on('data', (data: Buffer) => {
      collectedOutput.push(data.toString());
      // wait for local-registry to come online
      if (data.includes('http address')) {
        resolvedOrRejected = true;
        res(undefined);
      }
    });
    localRegistryProcess.stderr!.pipe(process.stderr);
    localRegistryProcess.on('error', (err: Error) => {
      console.error(collectedOutput.join(''));
      resolvedOrRejected = true;
      rej(err);
    });
  });
}

export function cleanUp(code: number): void {
  process.chdir(cwd);

  // try and terminate everything first
  try {
    if (!process.env.CI) {
      kill(0);
    }
    // eslint-disable-next-line no-empty
  } catch {}
  try {
    if (localRegistryProcess) localRegistryProcess.kill();
    // eslint-disable-next-line no-empty
  } catch {}
  // try killing everything after in case something hasn't terminated
  try {
    if (!process.env.CI) {
      kill(0, 'SIGKILL');
    }
    // eslint-disable-next-line no-empty
  } catch {}
  try {
    if (localRegistryProcess) localRegistryProcess.kill('SIGKILL');
    // eslint-disable-next-line no-empty
  } catch {}

  process.exit(code);
}

export async function publishPackagesToVerdaccio(): Promise<
  execa.ExecaChildProcess<string>
> {
  if (process.env.npm_config_registry?.indexOf('http://localhost') === -1) {
    throw Error(`
      ------------------
      ðŸ’£ ERROR ðŸ’£ => $NPM_REGISTRY does not look like a local registry'
      ------------------
    `);
  }

  const subprocess = execa('./publish-to-verdaccio.sh', [VERSION]);
  subprocess.stdout!.pipe(process.stdout);
  subprocess.stderr!.pipe(process.stderr);

  return await subprocess;
}

export async function runNpmInstall(): Promise<
  execa.ExecaChildProcess<string>
> {
  if (process.env.npm_config_registry?.indexOf('http://localhost') === -1) {
    throw Error(`
      ------------------
      ðŸ’£ ERROR ðŸ’£ => $NPM_REGISTRY does not look like a local registry'
      ------------------
    `);
  }

  const subprocess = execa('npm', ['install']);
  subprocess.stdout!.pipe(process.stdout);
  subprocess.stderr!.pipe(process.stderr);

  return await subprocess;
}

export async function runYarnInstall(): Promise<
  execa.ExecaChildProcess<string>
> {
  if (process.env.npm_config_registry?.indexOf('http://localhost') === -1) {
    throw Error(`
      ------------------
      ðŸ’£ ERROR ðŸ’£ => $NPM_REGISTRY does not look like a local registry'
      ------------------
    `);
  }

  const subprocess = execa('yarn', ['install']);
  subprocess.stdout!.pipe(process.stdout);
  subprocess.stderr!.pipe(process.stderr);

  return await subprocess;
}

export async function runNgAdd(): Promise<execa.ExecaChildProcess<string>> {
  if (process.env.npm_config_registry?.indexOf('http://localhost') === -1) {
    throw Error(`
      ------------------
      ðŸ’£ ERROR ðŸ’£ => $NPM_REGISTRY does not look like a local registry'
      ------------------
    `);
  }

  const subprocess = execa('npx', [
    'ng',
    'add',
    `@angular-eslint/schematics@${VERSION}`,
  ]);
  subprocess.stdout!.pipe(process.stdout);
  subprocess.stderr!.pipe(process.stderr);

  return await subprocess;
}

export async function runNgNew(
  workspaceName: string,
  createApplication = true,
): Promise<execa.ExecaChildProcess<string>> {
  if (process.env.npm_config_registry?.indexOf('http://localhost') === -1) {
    throw Error(`
      ------------------
      ðŸ’£ ERROR ðŸ’£ => $NPM_REGISTRY does not look like a local registry'
      ------------------
    `);
  }

  const ngNewArgs = [
    `--strict=true`,
    `--package-manager=npm`,
    `--interactive=false`,
  ];

  if (!createApplication) {
    ngNewArgs.push(`--create-application=false`);
  }

  const subprocess = execa('../../../node_modules/.bin/ng', [
    'new',
    ...ngNewArgs,
    workspaceName,
  ]);
  subprocess.stdout!.pipe(process.stdout);
  subprocess.stderr!.pipe(process.stderr);

  return await subprocess;
}

export async function runNgGenerate(
  args: string[],
): Promise<execa.ExecaChildProcess<string>> {
  if (process.env.npm_config_registry?.indexOf('http://localhost') === -1) {
    throw Error(`
      ------------------
      ðŸ’£ ERROR ðŸ’£ => $NPM_REGISTRY does not look like a local registry'
      ------------------
    `);
  }

  const subprocess = execa('npx', ['ng', 'generate', ...args]);
  subprocess.stdout!.pipe(process.stdout);
  subprocess.stderr!.pipe(process.stderr);

  return await subprocess;
}
