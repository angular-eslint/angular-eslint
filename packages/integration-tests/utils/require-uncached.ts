/**
 * It's important in our snapshot assertions which target modules on disk
 * that we are always grabbing a fresh copy of the file. Node's built-in
 * caching mechanism can get in the way of that so this utility helps
 * us work around the potential issue.
 *
 * (Having the return type as any makes it much easier to work with).
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function requireUncached(modulePath: string): any {
  delete require.cache[require.resolve(modulePath)];
  return require(modulePath);
}
