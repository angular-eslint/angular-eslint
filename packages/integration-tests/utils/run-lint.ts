/* eslint-disable @typescript-eslint/no-non-null-assertion */
import execa from 'execa';
import path from 'path';

import stripAnsi = require('strip-ansi');

const FIXTURES_DIR = path.join(__dirname, '../fixtures/');

function normalizeOutput(value: unknown): string {
  const ansiRemoved = stripAnsi(String(value));
  return (
    ansiRemoved
      .replace(
        new RegExp(`^${FIXTURES_DIR.replace(/\\/g, '\\\\')}(.*?)$`, 'gm'),
        (_, c1) => `__ROOT__/${c1.replace(/\\/g, '/')}`,
      )
      /**
       * This one is ridiculous...
       *
       * No matter what I try (including the use of `strip-ansi`), I do not seem to be
       * able to stop the inclusion of this extra `m` any other way (it is presumably
       * coming from an ANSI code becuase of its initial appearance in CI as [27m).
       * It happens locally and CI... but only exactly once.
       */
      .replace('mmax-len', 'max-len')
  );
}

export async function runLint(directory: string): Promise<string | undefined> {
  try {
    const subprocess = execa('npx', ['ng', 'lint'], {
      cwd: path.join(FIXTURES_DIR, directory),
    });

    /* eslint-disable @typescript-eslint/no-non-null-assertion */
    subprocess.stdout!.pipe(process.stdout);
    subprocess.stderr!.pipe(process.stderr);
    /* eslint-enable @typescript-eslint/no-non-null-assertion */

    const { stdout } = await subprocess;

    return normalizeOutput(stdout);
  } catch (error: any) {
    return normalizeOutput(error.stdout || error);
  }
}
