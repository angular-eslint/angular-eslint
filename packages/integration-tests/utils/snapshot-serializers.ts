function escapeRegExp(str: string) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function normalizeVersionOfPackage(str: string, pkg: string) {
  const regex = new RegExp(
    `("${escapeRegExp(pkg)}": "[~|^]\\d\\d?\\.)(\\d.\\d)"`,
  );
  return str.replace(regex, '$1X.X"');
}

const dependenciesToNormalize = [
  '@angular-devkit/build-angular',
  '@angular/cli',
  '@angular/compiler-cli',
  'ng-packagr',
  'typescript',
];

/**
 * Normalize dependencies controlled by the Angular CLI so that only the major version is explicitly
 * checked so that we can massively cut down on snapshot update noise across PRs.
 */
export const normalizeVersionsOfPackagesWeDoNotControl = {
  serialize(str: string) {
    for (const pkg of dependenciesToNormalize) {
      str = normalizeVersionOfPackage(str, pkg);
    }
    return str;
  },
  test(val: string) {
    // Only run if we think it's package.json contents
    return (
      val != null &&
      typeof val === 'string' &&
      dependenciesToNormalize.some((pkg) => val.includes(`"${pkg}"`))
    );
  },
};
