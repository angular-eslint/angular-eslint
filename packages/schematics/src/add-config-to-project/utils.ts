/**
 * Utils taken from various parts of Nx:
 * https://github.com/nrwl/nx
 *
 * Thanks, Nrwl folks!
 */
import { normalize } from '@angular-devkit/core';
import { Tree, SchematicContext, Rule } from '@angular-devkit/schematics';
import stripJsonComments from 'strip-json-comments';

/**
 * This method is specifically for reading JSON files in a Tree
 * @param host The host tree
 * @param path The path to the JSON file
 * @returns The JSON data in the file.
 */
function readJsonInTree<T = any>(host: Tree, path: string): T {
  if (!host.exists(path)) {
    throw new Error(`Cannot find ${path}`);
  }
  const contents = stripJsonComments(host.read(path)!.toString('utf-8'));
  try {
    return JSON.parse(contents);
  } catch (e) {
    throw new Error(`Cannot parse ${path}: ${e.message}`);
  }
}

function getWorkspacePath(host: Tree) {
  const possibleFiles = ['/workspace.json', '/angular.json', '/.angular.json'];
  return possibleFiles.filter((path) => host.exists(path))[0];
}

export function getProjectConfig(host: Tree, name: string): any {
  const workspaceJson = readJsonInTree(host, getWorkspacePath(host));
  const projectConfig = workspaceJson.projects[name];
  if (!projectConfig) {
    throw new Error(`Cannot find project '${name}'`);
  } else {
    return projectConfig;
  }
}

export function offsetFromRoot(fullPathToSourceDir: string): string {
  const parts = normalize(fullPathToSourceDir).split('/');
  let offset = '';
  for (let i = 0; i < parts.length; ++i) {
    offset += '../';
  }
  return offset;
}

function serializeJson(json: any): string {
  return `${JSON.stringify(json, null, 2)}\n`;
}

export function updateWorkspaceInTree<T = any, O = T>(
  callback: (json: T, context: SchematicContext, host: Tree) => O,
): Rule {
  return (host: Tree, context: SchematicContext): Tree => {
    const path = getWorkspacePath(host);
    host.overwrite(
      path,
      serializeJson(callback(readJsonInTree(host, path), context, host)),
    );
    return host;
  };
}
