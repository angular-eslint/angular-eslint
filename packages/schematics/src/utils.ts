/**
 * Some utils taken from various parts of Nx:
 * https://github.com/nrwl/nx
 *
 * Thanks, Nrwl folks!
 */
import type { Path } from '@angular-devkit/core';
import { join, normalize } from '@angular-devkit/core';
import type { Rule, SchematicContext, Tree } from '@angular-devkit/schematics';
import { callRule } from '@angular-devkit/schematics';
import type { Ignore } from 'ignore';
import ignore from 'ignore';
import stripJsonComments from 'strip-json-comments';

/**
 * This method is specifically for reading JSON files in a Tree
 * @param host The host tree
 * @param path The path to the JSON file
 * @returns The JSON data in the file.
 */
export function readJsonInTree<T = any>(host: Tree, path: string): T {
  if (!host.exists(path)) {
    throw new Error(`Cannot find ${path}`);
  }
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const contents = stripJsonComments(host.read(path)!.toString('utf-8'));
  try {
    return JSON.parse(contents);
  } catch (e) {
    throw new Error(`Cannot parse ${path}: ${e.message}`);
  }
}

/**
 * This method is specifically for updating JSON in a Tree
 * @param path Path of JSON file in the Tree
 * @param callback Manipulation of the JSON data
 * @returns A rule which updates a JSON file file in a Tree
 */
export function updateJsonInTree<T = any, O = T>(
  path: string,
  callback: (json: T, context: SchematicContext) => O,
): Rule {
  return (host: Tree, context: SchematicContext): Tree => {
    if (!host.exists(path)) {
      host.create(path, serializeJson(callback({} as T, context)));
      return host;
    }
    host.overwrite(
      path,
      serializeJson(callback(readJsonInTree(host, path), context)),
    );
    return host;
  };
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function getWorkspacePath(host: Tree) {
  const possibleFiles = ['/workspace.json', '/angular.json', '/.angular.json'];
  return possibleFiles.filter((path) => host.exists(path))[0];
}

type TargetsConfig = Record<string, { builder: string; options: unknown }>;

export function getTargetsConfigFromProject(
  projectConfig: { architect?: TargetsConfig } & { targets?: TargetsConfig },
): TargetsConfig | null {
  if (!projectConfig) {
    return null;
  }
  if (projectConfig.architect) {
    return projectConfig.architect;
  }
  // "targets" is an undocumented but supported alias of "architect"
  if (projectConfig.targets) {
    return projectConfig.targets;
  }
  return null;
}

export function isTSLintUsedInWorkspace(tree: Tree): boolean {
  const workspaceJson = readJsonInTree(tree, getWorkspacePath(tree));
  if (!workspaceJson) {
    return false;
  }

  for (const [, projectConfig] of Object.entries(
    workspaceJson.projects,
  ) as any) {
    const targetsConfig = getTargetsConfigFromProject(projectConfig);
    if (!targetsConfig) {
      continue;
    }

    for (const [, targetConfig] of Object.entries(targetsConfig)) {
      if (!targetConfig) {
        continue;
      }

      if (targetConfig.builder === '@angular-devkit/build-angular:tslint') {
        // Workspace is still using TSLint, exit early
        return true;
      }
    }
  }
  // If we got this far the user has no remaining TSLint usage
  return false;
}

export function getProjectConfig(host: Tree, name: string): any {
  const workspaceJson = readJsonInTree(host, getWorkspacePath(host));
  const projectConfig = workspaceJson.projects[name];
  if (!projectConfig) {
    throw new Error(`Cannot find project '${name}'`);
  } else {
    return projectConfig;
  }
}

export function offsetFromRoot(fullPathToSourceDir: string): string {
  const parts = normalize(fullPathToSourceDir).split('/');
  let offset = '';
  for (let i = 0; i < parts.length; ++i) {
    offset += '../';
  }
  return offset;
}

function serializeJson(json: any): string {
  return `${JSON.stringify(json, null, 2)}\n`;
}

export function updateWorkspaceInTree<T = any, O = T>(
  callback: (json: T, context: SchematicContext, host: Tree) => O,
): Rule {
  return (host: Tree, context: SchematicContext): Tree => {
    const path = getWorkspacePath(host);
    host.overwrite(
      path,
      serializeJson(callback(readJsonInTree(host, path), context, host)),
    );
    return host;
  };
}

export function addESLintTargetToProject(
  projectName: string,
  targetName: 'eslint' | 'lint',
): Rule {
  return updateWorkspaceInTree((workspaceJson) => {
    const existingProjectConfig = workspaceJson.projects[projectName];

    let lintFilePatternsRoot = '';

    // Default Angular CLI project at the root of the workspace
    if (existingProjectConfig.root === '') {
      lintFilePatternsRoot = 'src';
    } else {
      lintFilePatternsRoot = existingProjectConfig.root;
    }

    const eslintTargetConfig = {
      builder: '@angular-eslint/builder:lint',
      options: {
        lintFilePatterns: [
          `${lintFilePatternsRoot}/**/*.ts`,
          `${lintFilePatternsRoot}/**/*.html`,
        ],
      },
    };

    existingProjectConfig.architect[targetName] = eslintTargetConfig;

    return workspaceJson;
  });
}

/**
 * Utility to act on all files in a tree that are not ignored by git.
 */
export function visitNotIgnoredFiles(
  visitor: (file: Path, host: Tree, context: SchematicContext) => void | Rule,
  dir: Path = normalize(''),
): Rule {
  return (host, context) => {
    let ig: Ignore;
    if (host.exists('.gitignore')) {
      ig = ignore();
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      ig.add(host.read('.gitignore')!.toString());
    }

    function visit(_dir: Path) {
      if (_dir && ig?.ignores(_dir)) {
        return;
      }
      const dirEntry = host.getDir(_dir);
      dirEntry.subfiles.forEach((file) => {
        if (ig?.ignores(join(_dir, file))) {
          return;
        }
        const maybeRule = visitor(join(_dir, file), host, context);
        if (maybeRule) {
          callRule(maybeRule, host, context).subscribe();
        }
      });

      dirEntry.subdirs.forEach((subdir) => {
        visit(join(_dir, subdir));
      });
    }

    visit(dir);
  };
}

type ProjectType = 'application' | 'library';

export function setESLintProjectBasedOnProjectType(
  projectRoot: string,
  projectType: ProjectType,
  hasE2e?: boolean,
) {
  let project;
  if (projectType === 'application') {
    project = [
      `${projectRoot}/tsconfig.app.json`,
      `${projectRoot}/tsconfig.spec.json`,
    ];

    if (hasE2e) {
      project.push(`${projectRoot}/e2e/tsconfig.json`);
    }
  }
  // Libraries don't have an e2e directory
  if (projectType === 'library') {
    project = [
      `${projectRoot}/tsconfig.lib.json`,
      `${projectRoot}/tsconfig.spec.json`,
    ];
  }
  return project;
}

export function createRootESLintConfig(
  prefix: string | null,
  hasE2e?: boolean,
) {
  let codeRules;
  if (prefix) {
    codeRules = {
      '@angular-eslint/directive-selector': [
        'error',
        { type: 'attribute', prefix, style: 'camelCase' },
      ],
      '@angular-eslint/component-selector': [
        'error',
        { type: 'element', prefix, style: 'kebab-case' },
      ],
    };
  } else {
    codeRules = {};
  }

  return {
    root: true,
    ignorePatterns: ['projects/**/*'],
    overrides: [
      {
        files: ['*.ts'],
        parserOptions: {
          project: hasE2e
            ? ['tsconfig.json', 'e2e/tsconfig.json']
            : ['tsconfig.json'],
          createDefaultProgram: true,
        },
        extends: [
          'plugin:@angular-eslint/recommended',
          'plugin:@angular-eslint/template/process-inline-templates',
        ],
        rules: codeRules,
      },

      {
        files: ['*.html'],
        extends: ['plugin:@angular-eslint/template/recommended'],
        rules: {},
      },
    ],
  };
}

function createProjectESLintConfig(
  rootPath: string,
  projectRoot: string,
  projectType: ProjectType,
  prefix: string,
  hasE2e: boolean,
) {
  return {
    extends: `${offsetFromRoot(rootPath)}.eslintrc.json`,
    ignorePatterns: ['!**/*'],
    overrides: [
      {
        files: ['*.ts'],
        parserOptions: {
          project: setESLintProjectBasedOnProjectType(
            projectRoot,
            projectType,
            hasE2e,
          ),
          createDefaultProgram: true,
        },
        rules: {
          '@angular-eslint/directive-selector': [
            'error',
            { type: 'attribute', prefix, style: 'camelCase' },
          ],
          '@angular-eslint/component-selector': [
            'error',
            { type: 'element', prefix, style: 'kebab-case' },
          ],
        },
      },

      {
        files: ['*.html'],
        rules: {},
      },
    ],
  };
}

export function createESLintConfigForProject(projectName: string): Rule {
  return (tree: Tree) => {
    const angularJSON = readJsonInTree(tree, 'angular.json');
    const { root: projectRoot, projectType, prefix } = angularJSON.projects[
      projectName
    ];

    const hasE2e = determineTargetProjectHasE2E(angularJSON, projectName);

    /**
     * If the root is an empty string it must be the initial project created at the
     * root by the Angular CLI's workspace schematic
     */
    if (projectRoot === '') {
      return createRootESLintConfigFile(projectName);
    }
    return updateJsonInTree(
      join(normalize(projectRoot), '.eslintrc.json'),
      () =>
        createProjectESLintConfig(
          tree.root.path,
          projectRoot,
          projectType,
          prefix,
          hasE2e,
        ),
    );
  };
}

export function removeTSLintJSONForProject(projectName: string): Rule {
  return (tree: Tree) => {
    const angularJSON = readJsonInTree(tree, 'angular.json');
    const { root: projectRoot } = angularJSON.projects[projectName];
    const tslintJsonPath = join(normalize(projectRoot || '/'), 'tslint.json');
    if (tree.exists(tslintJsonPath)) {
      tree.delete(tslintJsonPath);
    }
  };
}

function createRootESLintConfigFile(projectName: string): Rule {
  return (tree) => {
    const angularJSON = readJsonInTree(tree, getWorkspacePath(tree));
    let lintPrefix: string | null = null;
    const hasE2e = determineTargetProjectHasE2E(angularJSON, projectName);

    if (angularJSON.projects?.[projectName]) {
      const { prefix } = angularJSON.projects[projectName];
      lintPrefix = prefix;
    }

    return updateJsonInTree('.eslintrc.json', () =>
      createRootESLintConfig(lintPrefix, hasE2e),
    );
  };
}

export function sortObjectByKeys(obj: Record<string, unknown>) {
  return Object.keys(obj)
    .sort()
    .reduce((result, key) => {
      return {
        ...result,
        [key]: obj[key],
      };
    }, {});
}

/**
 * To make certain schematic usage conversion more ergonomic, if the user does not specify a project
 * and only has a single project in their angular.json we will just go ahead and use that one.
 */
export function determineTargetProjectName(
  tree: Tree,
  maybeProject?: string,
): string | null {
  if (maybeProject) {
    return maybeProject;
  }
  const workspaceJson = readJsonInTree(tree, getWorkspacePath(tree));
  const projects = Object.keys(workspaceJson.projects);
  if (projects.length === 1) {
    return projects[0];
  }
  return null;
}

/**
 * Checking if the target project has e2e setup
 * Method will check if angular project architect has e2e configuration to determine if e2e setup
 */
export function determineTargetProjectHasE2E(
  angularJSON: any,
  projectName: string,
): boolean {
  return !!getTargetsConfigFromProject(angularJSON.projects[projectName])?.e2e;
}
