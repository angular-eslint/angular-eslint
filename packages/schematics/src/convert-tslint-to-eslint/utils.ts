import type { Rule, SchematicContext, Tree } from '@angular-devkit/schematics';
import { NodePackageInstallTask } from '@angular-devkit/schematics/tasks';
import * as assert from 'assert';
import { sortObjectByKeys } from '../utils';

export function updateArrPropAndRemoveDuplication(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  json: Record<string, any>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  configBeingExtended: Record<string, any>,
  arrPropName: string,
  deleteIfUltimatelyEmpty: boolean,
): void {
  json[arrPropName] = json[arrPropName] || [];
  configBeingExtended[arrPropName] = configBeingExtended[arrPropName] || [];
  json[arrPropName] = json[arrPropName].filter(
    (extended: string) => !configBeingExtended[arrPropName].includes(extended),
  );
  json[arrPropName] = Array.from(new Set(json[arrPropName]));
  if (deleteIfUltimatelyEmpty && json[arrPropName].length === 0) {
    delete json[arrPropName];
  }
}

export function updateObjPropAndRemoveDuplication(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  json: Record<string, any>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  configBeingExtended: Record<string, any>,
  objPropName: string,
  deleteIfUltimatelyEmpty: boolean,
): void {
  json[objPropName] = json[objPropName] || {};
  configBeingExtended[objPropName] = configBeingExtended[objPropName] || {};

  for (const [name, val] of Object.entries(json[objPropName])) {
    const valueOfSamePropInExtendedConfig =
      configBeingExtended[objPropName][name];

    try {
      assert.deepStrictEqual(val, valueOfSamePropInExtendedConfig);
      delete json[objPropName][name];
      // eslint-disable-next-line no-empty
    } catch {}
  }

  if (deleteIfUltimatelyEmpty && Object.keys(json[objPropName]).length === 0) {
    delete json[objPropName];
  }
}

export function ensureESLintPluginsAreInstalled(
  eslintPluginsToBeInstalled: string[],
): Rule {
  return (host: Tree, context: SchematicContext) => {
    if (!eslintPluginsToBeInstalled?.length) {
      return;
    }

    if (!host.exists('package.json')) {
      throw new Error(
        'Could not find a `package.json` file at the root of your workspace',
      );
    }

    const projectPackageJSON = (host.read('package.json') as Buffer).toString(
      'utf-8',
    );
    const json = JSON.parse(projectPackageJSON);
    json.devDependencies = json.devDependencies || {};

    const pluginsToInstall = [];

    for (const pluginName of eslintPluginsToBeInstalled) {
      if (
        !json.devDependencies[pluginName] &&
        !json.dependencies?.[pluginName]
      ) {
        json.devDependencies[pluginName] =
          // Hard code eslint-plugin-jsdoc version for now so that breaking rule change does not apply
          pluginName === 'eslint-plugin-jsdoc' ? '41.1.2' : 'latest';
        pluginsToInstall.push(pluginName);
      }
    }

    if (pluginsToInstall.length > 0) {
      context.logger.info(
        '\nINFO: To most closely match your tslint.json, the `latest` version of the following eslint plugin(s) have been installed:',
      );
      context.logger.info('\n  - ' + pluginsToInstall.join('\n  - '));
      context.logger.info(
        '\nPlease note, you may wish to pin these to a specific version number in your package.json, rather than leaving it open to `latest`.\n',
      );

      host.overwrite('package.json', JSON.stringify(json, null, 2));
      context.addTask(new NodePackageInstallTask());
    }

    return host;
  };
}

export function uninstallTSLintAndCodelyzer(): Rule {
  return (host: Tree, context: SchematicContext) => {
    if (!host.exists('package.json')) {
      throw new Error(
        'Could not find a `package.json` file at the root of your workspace',
      );
    }

    const projectPackageJSON = (host.read('package.json') as Buffer).toString(
      'utf-8',
    );
    const json = JSON.parse(projectPackageJSON);

    if (json.devDependencies) {
      delete json.devDependencies['tslint'];
      delete json.devDependencies['codelyzer'];
      json.devDependencies = sortObjectByKeys(json.devDependencies);
    }

    if (json.dependencies) {
      delete json.dependencies['tslint'];
      delete json.dependencies['codelyzer'];
      json.dependencies = sortObjectByKeys(json.dependencies);
    }

    host.overwrite('package.json', JSON.stringify(json, null, 2));
    context.addTask(new NodePackageInstallTask());

    return host;
  };
}
