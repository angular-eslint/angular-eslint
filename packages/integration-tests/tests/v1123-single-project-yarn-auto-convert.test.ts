/* eslint-disable @typescript-eslint/no-var-requires */
import path from 'path';
import {
  FIXTURES_DIR,
  LONG_TIMEOUT_MS,
  runConvertTSLintToESLint,
  runNgAdd,
  runYarnInstall,
} from '../utils/local-registry-process';
import { requireUncached } from '../utils/require-uncached';
import { runLint } from '../utils/run-lint';

const fixtureDirectory = 'v1123-single-project-yarn-auto-convert';

describe(fixtureDirectory, () => {
  jest.setTimeout(LONG_TIMEOUT_MS);

  beforeEach(async () => {
    process.chdir(path.join(FIXTURES_DIR, fixtureDirectory));
    await runYarnInstall();
    await runNgAdd();
    await runConvertTSLintToESLint([
      '--no-interactive',
      '--project',
      'v1123-single-project-yarn-auto-convert',
    ]);
  });

  it('it should pass linting after converting the out of the box Angular CLI setup (with a custom prefix set)', async () => {
    expect(
      requireUncached(
        '../fixtures/v1123-single-project-yarn-auto-convert/.eslintrc.json',
      ),
    ).toMatchSnapshot();

    expect(
      requireUncached(
        '../fixtures/v1123-single-project-yarn-auto-convert/angular.json',
      ).projects['v1123-single-project-yarn-auto-convert'].architect.lint,
    ).toMatchSnapshot();

    const lintOutput = await runLint(fixtureDirectory);
    expect(lintOutput).toMatchSnapshot();
  });
});
