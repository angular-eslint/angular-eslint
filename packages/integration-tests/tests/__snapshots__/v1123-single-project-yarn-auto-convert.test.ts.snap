// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v1123-single-project-yarn-auto-convert it should pass linting after converting the out of the box Angular CLI setup (with a custom prefix set) 1`] = `
Object {
  "ignorePatterns": Array [
    "projects/**/*",
  ],
  "overrides": Array [
    Object {
      "extends": Array [
        "plugin:@angular-eslint/ng-cli-compat",
        "plugin:@angular-eslint/ng-cli-compat--formatting-add-on",
        "plugin:@angular-eslint/template/process-inline-templates",
      ],
      "files": Array [
        "*.ts",
      ],
      "parserOptions": Object {
        "createDefaultProgram": true,
        "project": Array [
          "tsconfig.json",
          "e2e/tsconfig.json",
        ],
      },
      "rules": Object {
        "@angular-eslint/component-selector": Array [
          "error",
          Object {
            "prefix": "some-custom-prefix",
            "style": "kebab-case",
            "type": "element",
          },
        ],
        "@angular-eslint/directive-selector": Array [
          "error",
          Object {
            "prefix": "some-custom-prefix",
            "style": "camelCase",
            "type": "attribute",
          },
        ],
      },
    },
    Object {
      "extends": Array [
        "plugin:@angular-eslint/template/recommended",
      ],
      "files": Array [
        "*.html",
      ],
      "rules": Object {},
    },
  ],
  "root": true,
}
`;

exports[`v1123-single-project-yarn-auto-convert it should pass linting after converting the out of the box Angular CLI setup (with a custom prefix set) 2`] = `
Object {
  "builder": "@angular-eslint/builder:lint",
  "options": Object {
    "lintFilePatterns": Array [
      "src/**/*.ts",
      "src/**/*.html",
    ],
  },
}
`;

exports[`v1123-single-project-yarn-auto-convert it should pass linting after converting the out of the box Angular CLI setup (with a custom prefix set) 3`] = `
"
Linting \\"v1123-single-project-yarn-auto-convert\\"...

__ROOT__/v1123-single-project-yarn-auto-convert/src/test.ts
  13:5  error  Member call should be declared before all method definitions  @typescript-eslint/member-ordering

âœ– 1 problem (1 error, 0 warnings)
"
`;
